apps/opencs/view/render/terrainshapemode.cpp

```diff
@@ -0,0 +1,1459 @@
+    return false;
+}
+
+void CSVRender::TerrainShapeMode::drag (const QPoint& pos, int diffX, int diffY, double speedFactor)
+{
+    if (mDragMode == InteractionType_PrimaryEdit)
+    {
+        WorldspaceHitResult hit = getWorldspaceWidget().mousePick (pos, getWorldspaceWidget().getInteractionMask());
+        std::string cellId = getWorldspaceWidget().getCellId (hit.worldPos);
+        mTotalDiffY += diffY;
+        if (mIsEditing)
+        {
+            if (mShapeEditTool == ShapeEditTool_Drag) editTerrainShapeGrid(CSMWorld::CellCoordinates::toVertexCoords(mEditingPos), true);
+            if (mShapeEditTool != ShapeEditTool_Drag) editTerrainShapeGrid(CSMWorld::CellCoordinates::toVertexCoords(hit.worldPos), true);
+        }
+    }
+
+    if (mDragMode == InteractionType_PrimarySelect)
+    {
+        WorldspaceHitResult hit = getWorldspaceWidget().mousePick (pos, getWorldspaceWidget().getInteractionMask());
+        if (hit.hit && hit.tag == 0) selectTerrainShapes(CSMWorld::CellCoordinates::toVertexCoords(hit.worldPos), 0, true);
+    }
+
+    if (mDragMode == InteractionType_SecondarySelect)
+    {
+        WorldspaceHitResult hit = getWorldspaceWidget().mousePick (pos, getWorldspaceWidget().getInteractionMask());
+        if (hit.hit && hit.tag == 0) selectTerrainShapes(CSMWorld::CellCoordinates::toVertexCoords(hit.worldPos), 1, true);
+    }
+}
+
+void CSVRender::TerrainShapeMode::dragCompleted(const QPoint& pos)
+{
+    if (mDragMode == InteractionType_PrimaryEdit)
+    {
+        if (mIsEditing)
+        {
+            mTotalDiffY = 0;
+            mIsEditing = false;
+        }
+
+        applyTerrainEditChanges();
+
+        if (CSVRender::PagedWorldspaceWidget *paged = dynamic_cast<CSVRender::PagedWorldspaceWidget *> (&getWorldspaceWidget()))
+            paged->resetAllAlteredHeights();
+    }
+}
+
+
+void CSVRender::TerrainShapeMode::dragAborted()
+{
+    if (CSVRender::PagedWorldspaceWidget *paged =
+        dynamic_cast<CSVRender::PagedWorldspaceWidget *> (&getWorldspaceWidget()))
+    {
+        paged->resetAllAlteredHeights();
+        mTotalDiffY = 0;
+    }
+}
+
+void CSVRender::TerrainShapeMode::dragWheel (int diff, double speedFactor)
+{
+}
+
+void CSVRender::TerrainShapeMode::sortAndLimitAlteredCells()
+{
+    bool passing = false;
+    int passes = 0;
+
+    std::sort(mAlteredCells.begin(), mAlteredCells.end());
+    mAlteredCells.erase(std::unique(mAlteredCells.begin(), mAlteredCells.end()), mAlteredCells.end());
+
+    while (!passing) // Multiple passes are needed when steepness problems arise for both x and y axis simultaneously
+    {
+        passing = true;
+        for(CSMWorld::CellCoordinates cellCoordinates: mAlteredCells)
+        {
+            limitAlteredHeights(cellCoordinates);
+        }
+        std::reverse(mAlteredCells.begin(), mAlteredCells.end()); //Instead of alphabetical order, this should be fixed to sort cells by cell coordinates
```

Comment:
Is final order after this loop end is important for `mAlteredCells`?

Reply:
I haven't tested removing it, but in theory, it is important. I'm not certain the land limiting function `limitAlteredHeights(cellCoordinates);` takes proper use of it right now, as the sorting is alphabetical `std::string`s instead of sorting by X and Y coordinates. Land limiting should always check and limit all cells from left to right, and up to down, then backwards. By doing it this way, land limiting should continue it's work right from the cell edge.

If you imagine a grid of cells, and put the limiting algorithm to go through them in random order, it's likely that the limiting function will change the heights at the cell edges, and there might be cases when the limiting function will test the heights at a place that it has already changed, and this might have unexpected results.